server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -0,0 +1,741 @@
+    }
+
+    /**
+     * Remove all persisted cluster states from the given data paths, for use in tests. Should only be called when there is no open
+     * {@link Writer} on these paths.
+     */
+    public static void deleteAll(Path[] dataPaths) throws IOException {
+        for (Path dataPath : dataPaths) {
+            Lucene.cleanLuceneIndex(new SimpleFSDirectory(dataPath.resolve(METADATA_DIRECTORY_NAME)));
+        }
+    }
+
+    // exposed for tests
+    Directory createDirectory(Path path) throws IOException {
+        // it is possible to disable the use of MMapDirectory for indices, and it may be surprising to users that have done so if we still
+        // use a MMapDirectory here, which might happen with FSDirectory.open(path). Concurrency is of no concern here so a
+        // SimpleFSDirectory is fine:
+        return new SimpleFSDirectory(path);
+    }
+
+    public Path[] getDataPaths() {
+        return dataPaths;
+    }
+
+    public static class OnDiskState {
+        private static final OnDiskState NO_ON_DISK_STATE = new OnDiskState(null, null, 0L, 0L, Metadata.EMPTY_METADATA);
+
+        private final String nodeId;
+        private final Path dataPath;
+        public final long currentTerm;
+        public final long lastAcceptedVersion;
+        public final Metadata metadata;
+
+        private OnDiskState(String nodeId, Path dataPath, long currentTerm, long lastAcceptedVersion, Metadata metadata) {
+            this.nodeId = nodeId;
+            this.dataPath = dataPath;
+            this.currentTerm = currentTerm;
+            this.lastAcceptedVersion = lastAcceptedVersion;
+            this.metadata = metadata;
+        }
+
+        public boolean empty() {
+            return this == NO_ON_DISK_STATE;
+        }
+    }
+
+    /**
+     * Returns the node metadata for the given data paths, and checks if the node ids are unique
+     * @param dataPaths the data paths to scan
+     */
+    @Nullable
+    public static NodeMetadata nodeMetadata(Path... dataPaths) throws IOException {
+        String nodeId = null;
+        for (final Path dataPath : dataPaths) {
+            final Path indexPath = dataPath.resolve(METADATA_DIRECTORY_NAME);
+            if (Files.exists(indexPath)) {
+                try (DirectoryReader reader = DirectoryReader.open(new SimpleFSDirectory(dataPath.resolve(METADATA_DIRECTORY_NAME)))) {
+                    final Map<String, String> userData = reader.getIndexCommit().getUserData();
+                    assert userData.get(NODE_VERSION_KEY) != null;
+
+                    final String thisNodeId = userData.get(NODE_ID_KEY);
+                    assert thisNodeId != null;
+                    if (nodeId != null && nodeId.equals(thisNodeId) == false) {
+                        //Do nothing, because the metadata does not belong to this node
+                    } else if (nodeId == null) {
+                        nodeId = thisNodeId;
```

Comment:
ES also has logic to override the version field, why is that not relevant for us?

Reply:
We don't have the version yet in our metadata.

Reply:
Maybe we should backport this before.

Reply:
I think it's also okay to follow up on it if it is not relevant to the persistance change itself.